### ISSUE TITLE
Spurious flow errors

### ISSUE THREAD

**Issue by CheshireSwift** [dense-analysis/ale#1582]
## Information

**VIM version**

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jul 26 2017 19:10:24)
Included patches: 1-503, 505-642
```

Operating System: macOS High Sierra 10.13.3

### :ALEInfo

```
 Current Filetype: javascript
Available Linters: ['eslint', 'flow', 'jscs', 'jshint', 'standard', 'xo']
  Enabled Linters: ['eslint', 'flow']
 Linter Variables:
let g:ale_javascript_eslint_executable = 'eslint'
let g:ale_javascript_eslint_options = ''
let g:ale_javascript_eslint_suppress_eslintignore = 0
let g:ale_javascript_eslint_suppress_missing_config = 0
let g:ale_javascript_eslint_use_global = 0
let g:ale_javascript_flow_executable = 'flow'
let g:ale_javascript_flow_use_global = 0
let g:ale_javascript_flow_use_home_config = 0
let g:ale_javascript_flow_use_respect_pragma = 1
 Global Variables:
let g:ale_cache_executable_check_failures = 0
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = 50
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'javascript': ['importjs', 'trim_whitespace', 'remove_trailing_lines', 'eslint']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_lint_on_insert_leave = 0
let g:ale_linter_aliases = {}
let g:ale_linters = {'javascript': ['eslint', 'flow']}
let g:ale_linters_explicit = 0
let g:ale_list_window_size = 10
let g:ale_list_vertical = 0
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = 0
let g:ale_open_list = 0
let g:ale_pattern_options = {}
let g:ale_pattern_options_enabled = 0
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = 'X'
let g:ale_sign_info = '?'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = 'X'
let g:ale_sign_style_warning = '?'
let g:ale_sign_warning = '?'
let g:ale_statusline_format = ['X %d', '? %d', '']
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:
(executable check - failure) eslint
(executable check - success) /Users/stc/aletest/node_modules/.bin/flow
(started) ['/bin/zsh', '-c', '''/Users/stc/aletest/node_modules/.bin/flow'' --version; echo']
(executable check - failure) eslint
(finished - exit code 0) ['/bin/zsh', '-c', '''/Users/stc/aletest/node_modules/.bin/flow'' --version; echo']
(finished - exit code 0) ['/bin/zsh', '-c', '''/Users/stc/aletest/node_modules/.bin/flow'' check-contents --respect-pragma --json --from ale ''/Users/stc/aletest/Foo.js'' < ''/var/folders/bf/vd8kydz11dzcc6l8ysv2wwrxdvdn2b/T/vq8bEm7/1/F
oo.js''; echo']
<<<OUTPUT STARTS>>>
{"flowVersion":"0.72.0","jsonVersion":"1","errors":[{"kind":"infer","level":"error","suppressions":[],"extra":[{"message":[{"context":null,"descr":"References:","type":"Blame","path":"","line":0,"endline":0,"start":1,"end":0}]},{"messa
ge":[{"context":"class Foo {","descr":"[1]","type":"Blame","loc":{"source":"/Users/stc/aletest/Foo.js","type":"SourceFile","start":{"line":4,"column":7,"offset":43},"end":{"line":4,"column":9,"offset":46}},"path":"/Users/stc/aletest/Fo
o.js","line":4,"endline":4,"start":7,"end":9}]},{"message":[{"context":"export default function bar(foo: Foo) {}","descr":"[2]","type":"Blame","loc":{"source":"/Users/stc/aletest/bar.js","type":"SourceFile","start":{"line":4,"column":3
4,"offset":70},"end":{"line":4,"column":36,"offset":73}},"path":"/Users/stc/aletest/bar.js","line":4,"endline":4,"start":34,"end":36}]}],"message":[{"context":"    bar(this)","descr":"Cannot call `bar` with `this` bound to `foo` becaus
e `Foo` [1] is incompatible with `Foo` [2].","type":"Blame","loc":{"source":"/Users/stc/aletest/Foo.js","type":"SourceFile","start":{"line":6,"column":9,"offset":75},"end":{"line":6,"column":12,"offset":79}},"path":"/Users/stc/aletest/
Foo.js","line":6,"endline":6,"start":9,"end":12}]}],"passed":false}
<<<OUTPUT ENDS>>>
```

## What went wrong

ALE consistently has "Cannot call `bar` with `this` bound to `foo` becaus
e `Foo` [1] is incompatible with `Foo` [2]." error, despite flow (global or explicit local) reporting no errors.

Also tried replicating the command run by ALE, still no errors:
```
[ 1621 ][ stc@MacBook-Pro-2:~/aletest ]
$ /bin/zsh -c /Users/stc/aletest/node_modules/.bin/flow check-contents --respect-pragma --json --from ale /Users/stc/aletest/Unit.js
No errors!
```

## Reproducing the bug

Steps for repeating the bug:

1. Create the following files:

Foo.js:
```javascript
/* @flow */
import bar from './bar'

class Foo {
  fooDoStuff()  {
    bar(this)
  }
}
export default Foo
```

bar.js:
```javascript
/* @flow */
import Foo from './Foo'

export default function bar(foo: Foo) {}
```

2. Open Foo.js.
3. Specified error is output.

**Comment by TrySound** [comment]
By the way the latest (0.72) version introduced lsp support.

**Comment by hsanson** [comment]
Flow is being deprecated #3859 #2649
